text/0106-artifact-indices.md

```diff
@@ -0,0 +1,175 @@
+this is based on release bundles and artifact indices which are in the process of being phased out.
+
+- **pro**: Would allow querying / accessing the needed files.
+- **con**: Requires downloading / querying the index on API servers.
+- **con**: Based on deprecated technology that is in the process of being phased out.
+
+## Using a replacement flat-file index
+
+This would involve creating a replacement for artifact indexes based on the newer artifact bundles. Instead of a `JSON`
+index, a flat-file format was proposed. However the exact format is rather an insignificant detail.
+
+- **pro**: Would allow querying / accessing the needed files.
+- **pro**: Would solve one pain-point of artifact indexes: No more `Release` object needed.
+- **con**: Requires downloading / querying the index on API servers.
+- **con**: Needs to re-create / replicate most of the artifact-index logic, and inherits all of its cons:
+  - updating / merging the index on uploads
+  - the index growing monotonically on disjoint uploads
+
+## Index individual files in the Database
+
+This would create a similar database table / index to `ReleaseFile`, but with different constraints.
+
+- **pro**: Files individually indexed.
+- **pro**: A database-based solution might be simpler than maintaining a separate index manually.
+- **pro**: No `Release` object needed.
+- **pro**: No need to download / process any index on API servers.
+- **pro**: Would work well with auto-expiration / partitioning if implemented correctly.
+- **con**: High load on Database and its indices.
+- **con**: Might require a custom solution to better compress `release` and `url`, possibly by using some kind of interning.
+
+The table could look something the following:
+
+- org / project
+- (optional) release + dist
+- url
+- date_added / touched
+- artifact bundle id
+
+## Hybrid Bundles / Individual files
+
+The idea was brought up to only index individual files if the number of uploaded bundles per release exceeds a certain
+threshold `N`.
+
+This takes advantage of the fact that the downstream consumer (Symbolicator) will query all the returned `N` bundles
+for the needed files as well.
+
+- **mixed**: All the pros/cons of the above _individual files_, just with a lower number of indexed files.
+- **pro**: No additional overhead for the one-bundle-per-release use-case.
+- **con**: Needs jobs and logic to backfill a per-file index if the threshold was exceeded.
+
+## _Store_ individual files
+
+Instead of just _indexing_ files individually and still referring to the bundle they are included in, it might be possible
+to _store_ files themselves individually, possibly deduplicated by a content hash.
+
+- **pro**: Solves file decompression performance of Symbolicator.
+- **pro**: We _assume_ that most files inside of bundles are unused.
+- **pro**: We _assume_ that a lot of files will not change across bundles (polyfills, dependencies, etc).
+- **con**: We need a better way to store many small files.
```

Comment:
Better in what way? Is it too expensive, too hard to organize, or something else?

Reply:
Lets see if @mitsuhiko wants to chime in here, as he was very vocal against doing this. I believe the reason is that depending on storage vendor, you are billed by number of files, which can get extremely expensive storing a ton of small files.